File: graph/802/802.py
class Solution:

    def eventualSafeNodes(self, graph):
        n = len(graph)
        visited = [False]*n
        instack = [False]*n
        result = []

        for i in range(n):
            self.dfs(i, graph, visited, instack)

        for i in range(n):
            if not instack[i]:
                result.append(i)

        return result
    

    def dfs(self, node, graph, visited, instack):
        if instack[node]:
            return True

        if visited[node]:
            return False
        
        instack[node] = True
        visited[node] = True

        for n in graph[node]:
            if self.dfs(n, graph, visited, instack):
                return True
            
        instack[node] = False
        
        return False

File: graph/802/802.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <queue>
using namespace std;

class Solution {
    public:
        vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
            int n=graph.size();
            vector<int> visited(n), instack(n);
            vector<int> result;

            for(int i=0; i<n; i++){
                dfs(i, graph, visited, instack);
            }

            // for(int i=0; i<n; i++){
            //     cout<<instack[i]<<"";
            // }

            for(int i=0; i<n; i++){
                if(!instack[i]) result.push_back(i);
            }
            return result;
        }

        bool dfs(int node, vector<vector<int>>& graph, vector<int>& visited, vector<int>& instack) {
            if(instack[node]) return true;
            if(visited[node]) return false;
            
            instack[node]=true;
            visited[node]=true;

            for(int n:graph[node]){
                if(dfs(n, graph, visited, instack)) return true;
            }
            instack[node] = false;
            return false;
        }
};


int main() {
    Solution s;
    vector<vector<int>> graph = {{1,2},{2,3},{5},{0},{5},{},{}};
    
    auto result = s.eventualSafeNodes(graph);

    for(int n:result){
        cout<<n<<" ";
    }

    return 0;
}